### ISSUE TITLE
Make header checks in dropshot::test_util configurable.

### ISSUE THREAD

**Issue by arun-koshy** [oxidecomputer/dropshot#299]
Context: We require setting [CORS headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) as part of our Rest Server implementation. We have started a discussion about it in another [issue](https://github.com/oxidecomputer/dropshot/issues/57#issuecomment-1063721621) on this repo.

We have set the response headers but it causes our unit tests to fail because the test_util is setup to [panic](https://docs.rs/dropshot/latest/src/dropshot/test_util.rs.html#212) if any extra headers are added to the response. Can we make this configurable?

**Comment by ahl** [comment]
I did not realize that `test_util` was pub. @davepacheco is that intentional?

**Comment by davepacheco** [comment]
Yes.  `test_util` was for two purposes: (1) for testing dropshot itself, and (2) to help consumers test their Dropshot APIs.  In Omicron, we were using this pretty heavily until pretty recently and I believe we're still using it a bunch for internal APIs.  We started needing headers for authn and found that the interface was getting pretty clunky with all the super specific functions that are just slight variants of one another.  See discussion in #165.  We made this builder-based interface instead:
https://github.com/oxidecomputer/omicron/blob/main/nexus/test-utils/src/http_testing.rs#L17-L53

I think it's strictly better than what's in `dropshot::test_util` today, though it too can be a bit unwieldy, as you need an `await` and two `unwrap`s at every call site that wants to verify a response body (because it no longer asserts inside, but returns an error instead).  It may also be worth looking at `reqwest` instead, which has a reasonably ergonomic interface for stuff like this.

Anyway, the list of restricted headers that's there is really oriented around testing dropshot itself.  It makes sense to make this configurable for people using it to test their own APIs.  I think that's quite a lot easier to do with the builder-based interface that I linked above: you could easily add a `.allow_response_header()` there.

Options I see:

- move that builder interface into Dropshot or even a standalone crate and augment it with the ability to tweak allowed headers
- use reqwest instead
- add some new config interface for the existing `dropshot::test_util` thing (I don't like this, since we're moving away from what's in `dropshot::test_util` today)

Thoughts?

**Comment by jayhale** [comment]
@davepacheco +1 for moving the builder into Dropshot.  It has served all of the needs for an authenticated + rate-limited restish API I recently built with Dropshot.

I found this after realizing that the tools in `test_util` don't allow setting headers on the request, e.g., for authentication.
